text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,458 @@
+  // No need to worry or care about where this is called, this will _always_ work!
+  Sentry.getGlobalScope().setTag("revision", "abc");
+}
+```
+
+The global scope is _not_ the initial scope, but a separate, special scope. The global scope is different for each client, in order to avoid polluting this on accident when working with multiple completely separate clients.
+
+You can get the current global scope via `getGlobalScope()`. There _may_ be a function `setGlobalScope(scope)` to replace the global scope - or SDKs can decide that there is no need to replace the global scope, you can only mutate it.
+
+If you call `getGlobalScope()` before a client is initialized, we should still get a global scope back (tied to a Noop client). Once an actual client is initialized, the global scope of the noop client should be merged into the new global scope for the new client. This should ensure that even if you call `getGlobalScope().setTag(...)` before the SDK is initialized, no data is lost:
+
+```js
+// This global scope has a Noop client, because init was not called yet!
+const globalScopeBeforeInit = Sentry.getGlobalScope();
+globalScopeBeforeInit.setTag("tag1", "aa");
+
+// When init() is called, we make sure to inherit from the global scope, if data was set before
+Sentry.init();
+
+const globalScope = Sentry.getGlobalScope();
+// A new scope instance is created as global scope when a client is created & bound to current scope
+globalScope !== globalScopeBeforeInit;
+// Global scope inherits from globalScopeBeforeInit
+globalScope.getTags(); // -> tag1: aa
+```
+
+The reason that the global scope is not the same as the initial scope of a client, is that you cannot accidentally mutate it - nothing ever inherits off the global scope.
+
+### How to document global scopes
+
+Global scopes will be documented in a new section under "Enriching Events > Global Data", which will explain how to use `getGlobalScope()` and when/how data added there will be applied.
+
+### Isolation Scope
+
+Furthermore, there are also **Isolation Scopes**.
+
+An isolation scope is similar to a global scope, but tied to a specific section of your app/code that should be isolated.
+In a server side language, an isolation scope will commonly be created per request. Each request gets an isolated scope that you can assign data to, which should be applied to all events that are processed in this request. However, an isolation scope is not limited to requests - you could also have an isolation scope for a scheduled job, or for a zone in your UI (in theory).
+
+The isolation scope can be used to e.g. add user or request data that should be applied to any event for the current request (or similar isolation unit). Similar to the global scope, it does not matter where you call it, but you can always be sure that data is correctly attached/available. This can especially be useful when writing hooks or plugins where you may not easily get access to the scope of a request:
```

Comment:
Let's explicitly call out that breadcrumbs should be added on the isolation scope.

Reply:
I mention this further down here: https://github.com/getsentry/rfcs/blob/fn/merge-hub-scope/text/0122-sdk-hub-scope-merge.md#what-should-be-called-from-top-level-methods

but it's a general challenge for this RFC that stuff feels like it should be explained earlier that comes up later, making it hard to keep it somewhat readable 😬

Reply:
My comment here was I think that the breadcrumbs use case helps build the best intuition here why we need a isolation scope concept, it's also why we need something like this while opentelemetry does not fwiw.